compiler/compile/OMRCompilation.cpp

```diff
@@ -440,21 +440,25 @@ OMR::Compilation::Compilation(
       _options->setOption(TR_EnableOSROnGuardFailure, false);
       }
 
-   if (_options->getOption(TR_EnableHCR))
+   // Do not default OSR and HCR on if:
+   // NextGenHCR is disabled, as they are enabled for it
+   // OSR is explicitly disabled
+   // FSD is enabled, as HCR cannot be enabled with it
+   //
+   // Defaults are performed this late to account for the disable OSR above, as we do not want to enable HCR in cases where
+   // its not necessary
+   if (!self()->getOption(TR_DisableNextGenHCR) && !self()->getOption(TR_DisableOSR) && !self()->getOption(TR_FullSpeedDebug))
       {
-      if (!self()->getOption(TR_DisableOSR))
-         self()->setOption(TR_EnableOSR); // Make OSR the default for HCR
-      if (_options->getOption(TR_EnableNextGenHCR))
-         {
-         self()->setOption(TR_EnableOSR);
-         _options->setOption(TR_DisableOSR, false);
-         }
+      self()->setOption(TR_EnableOSR); // OSR and HCR must be enabled for NextGenHCR
+      self()->setOption(TR_EnableHCR);
       }
 
    if (_options->getOption(TR_EnableOSR))
       {
       // Current implementation of partial inlining will break OSR
       self()->setOption(TR_DisablePartialInlining);
+      // Can't yet transition a MH thunk frame into equivalent interpreter frames
+      self()->setOption(TR_DisableMethodHandleThunks);
```

Comment:
This isn't the right answer - when compiling a MH thunk we want to disable OSR - we cannot OSR out of the top most method frame (ie method -1) since it is a thunk with no interpreter equivalent. That doesn't mean we can't support HCR since thunks themselves are immune from redefinition. We use traditional HCR guards for all the 'normal' bodies inlined into the thunk. For thunks inlined into thunks and similar the redefition of the thunk archetype only changes the code template and so future compiles will pickup the new bytecodes, but we do not compensate and cause existing thunks to pickup the new code - per discussion with Dan.

This needs to be revised since it will negatively affect performance downstream from here.